assets: Add `wp_register_script` wrapper (and then use it everywhere)

anomiex opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
Our Webpack-built scripts tend to need a bunch of boilerplate:
 - A call to `Assets::get_file_url_for_environment()` for possible
   debugging.
 - A call to `wp_register_style()` for extracted CSS. Possibly with
   detection of RTL.
 - Loading of dependencies and version provided by
   `@wordpress/dependency-extraction-webpack-plugin`.
 - Avoiding WPCom's broken minifier.
    
The wrapper handles all of that.

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
PT: p9dueE-3MG-p2

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Make sure things all still work.

anomiex committed:
assets: Add `wp_register_script` wrapper

Our Webpack-built scripts tend to need a bunch of boilerplate:
 - A call to `Assets::get_file_url_for_environment()` for possible
   debugging.
 - A call to `wp_register_style()` for extracted CSS. Possibly with
   detection of RTL.
 - Loading of dependencies and version provided by
   `@wordpress/dependency-extraction-webpack-plugin`.
 - Avoiding WPCom's broken minifier.

The wrapper handles all of that.

anomiex committed:
Use it in Lazy-Images

anomiex committed:
Use it in Connection-UI

anomiex committed:
Use it in the Backup plugin

anomiex committed:
Use it in Jetpack

Jetpack has a bunch of different ways scripts get built. This only
touches the ones that are built with our webpack config.

Further, this doesn't touch admin.js or the blocks' view.js. For the
former, the JS and CSS are being registered in different places and the
CSS is being enqueued twice for some reason. For the latter, there's a
bunch of extra logic for handling previews that I didn't want to mess
with in this PR.

anomiex committed:
Version bumps

anomiex commented:
A few items of note regarding the package updates:

- `wp_set_script_translations` only works when the handle has been registered. A few scripts were calling it with other handles; I removed these.
- WordPress's 'rtl' style data doesn't fit with how we've been building our rtl.css. Either you set it to a URL where it assumes the rtl.css augments the existing .css (ours replaces it), or you set it to "replace" which assumes the file is named "-rtl.css" instead of `.rtl.css`. So I didn't bother trying to keep any of these.
- I also didn't bother trying to keep logic around extracted deps when `.asset.php` is missing. Most of the time if `.asset.php` is missing then the JS is missing too. If `SCRIPT_DEBUG` and `nonmin_path` are both set, the non-bundled JS loaded should be able to import its deps itself.
- Several places in Jetpack were ignoring the `.asset.php` version in favor of `JETPACK_VERSION`. That seems pointless, if we have the version we should use it.

anomiex committed:
More version bumps?

anomiex committed:
Drop the "once" conditions so it reports what it was actually called with

anomiex committed:
Don't use the full path in the filemtime hack

It can vary based on the environment, duh.

anomiex committed:
Make the mocked `add_query_arg` work more like the foot-gun the function really is

kangzj reviewed (dismissed):
Thanks for tackling this! It  works fine and the changes look reasonable to me.

jeherve commented on projects/packages/connection-ui/src/class-admin.php:
```diff
@@ -60,14 +62,16 @@ public function register_submenu_page() {
-
-			wp_enqueue_style( 'jetpack_connection_ui_style', plugin_dir_url( __DIR__ ) . 'build/index.css', array( 'wp-components' ), $build_assets['version'] );
-			wp_style_add_data( 'jetpack_connection_ui_style', 'rtl', plugin_dir_url( __DIR__ ) . 'build/index.rtl.css' );
+			Assets::register_script(
+				'jetpack_connection_ui',
+				'../build/index.js',
+				__FILE__,
+				array(
+					'in_footer' => true,
+				)
+			);
+			Assets::enqueue_script( 'jetpack_connection_ui' );
+			wp_add_inline_script( 'jetpack_connection_ui', $this->get_initial_state(), 'before' );
```
Should we keep the translation handling in here?

```suggestion
			Assets::register_script(
				'jetpack_connection_ui',
				'../build/index.js',
				__FILE__,
				array(
					'in_footer' => true,
				)
			);
			Assets::enqueue_script( 'jetpack_connection_ui' );
			wp_add_inline_script( 'jetpack_connection_ui', $this->get_initial_state(), 'before' );
			wp_set_script_translations( 'jetpack_connection_ui', 'jetpack' );
```

anomiex replied:
Should we fix the broken translation handling instead of deleting it? Sure, we can do that.

If my idea at https://github.com/Automattic/jetpack/issues/21690#issuecomment-966621370 pans out, I may wind up making "textdomain" an option to `Assets::register_script()` and have it complain if the script depends on wp-i18n but doesn't specify a domain.

jeherve commented on projects/plugins/backup/src/php/class-jetpack-backup.php:
```diff
@@ -82,36 +83,15 @@ public function admin_init() {
-			'rtl',
-			plugins_url( 'build/index.rtl.css', JETPACK_BACKUP_PLUGIN_ROOT_FILE )
-		);
+		wp_add_inline_script( 'jetpack-backup', $this->get_initial_state(), 'before' );
```
Same here? Actually this may have been broken, since the script handle wasn't correct. 

```suggestion
		wp_add_inline_script( 'jetpack-backup', $this->get_initial_state(), 'before' );
		wp_set_script_translations( 'jetpack-backup', 'jetpack-backup' );
```

anomiex replied:
Again, it's more "fix" than "keep".

jeherve commented on projects/packages/assets/src/class-assets.php:
```diff
@@ -223,4 +223,171 @@ public static function staticize_subdomain( $url ) {
+
+			$url = self::normalize_path( plugins_url( $csspath, $relative_to ) );
+			if ( null !== $options['minify'] ) {
+				$url = add_query_arg( 'minify', $options['minify'] ? 'true' : 'false', $url );
```
Given that the `minify` string is only useful on WordPress.com right now (I think?), what would you think about only adding `minify=false` when on WordPress.com simple, via the `IS_WPCOM` constant?

anomiex replied:
Personally I'd rather be consistent and always add it. It shouldn't hurt anything to include it.

anomiex committed:
Update projects/packages/connection-ui/src/class-admin.php

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

anomiex committed:
Merge remote-tracking branch 'origin/master' into add/assets-wp_register_script-wrapper

anomiex committed:
Update projects/plugins/backup/src/php/class-jetpack-backup.php

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

anomiex committed:
Use the new stuff for the instant-search main bundle too

Somehow I missed it the first time around. Plus, we need to fix a bug in
the forked AddReadableJsAssetsWebpackPlugin.

anomiex commented:
r235089-wpcom
